bayes_kit/drghmc.py

```diff
@@ -0,0 +1,471 @@
+        if not (proposals >= 1):
+            raise ValueError(
+                f"proposals must be greater than or equal to 1, not {proposals}"
+            )
+        return proposals
+
+    def _validate_leapfrog_stepsizes(
+        self, leapfrog_stepsizes: list[float]
+    ) -> list[float]:
+        """Validate list of leapfrog stepsizes.
+
+        Args:
+            leapfrog_stepsizes: list of non-increasing leapfrog stepsizes
+
+        Raises:
+            TypeError: leapfrog stepsizes is not a list
+            ValueError: leapfrog stepsize list is of incorrect length
+            TypeError: leapfrog stepsize list contains non-float stepsizes
+            ValueError: leapfrog stepsize list contains non-positive stepsizes
+            ValueError: leapfrog stepsize list contains decreasing stepsizes
+
+        Returns:
+            list of validated leapfrog stepsizes
+        """
+        if not (type(leapfrog_stepsizes) is list):
+            raise TypeError(
+                f"leapfrog_stepsizes must be a list, not {type(leapfrog_stepsizes)}"
+            )
+        if len(leapfrog_stepsizes) != self._proposals:
+            raise ValueError(
+                f"leapfrog_stepsizes must be a list of length {self._proposals}, not"
+                f" length {len(leapfrog_stepsizes)}, so that each proposal has a "
+                "specified leapfrog stepsize"
+            )
+        for idx, stepsize in enumerate(leapfrog_stepsizes):
+            if not (type(stepsize) is float):
+                raise TypeError(
+                    f"leapfrog stepsizes must be of type float, not {type(stepsize)} "
+                    f"at index {idx}"
+                )
+            if not stepsize > 0:
+                raise ValueError(
+                    f"leapfrog stepsizes must be positive, but found stepsize of "
+                    f"{stepsize} at index {idx}"
+                )
+        for idx, (prev, cur) in enumerate(
+            zip(leapfrog_stepsizes[:-1], leapfrog_stepsizes[1:])
+        ):
+            if not cur <= prev:
+                raise ValueError(
+                    f"leapfrog stepsizes must be non-increasing, but found stepsize of "
+                    f"{cur} at index {idx + 1} which is greater than stepsize of "
+                    f"{prev} at index {idx}"
+                )
+        return leapfrog_stepsizes
+
+    def _validate_leapfrog_stepcounts(
+        self, leapfrog_stepcounts: list[int]
+    ) -> list[int]:
+        """Validate list of leapfrog stepcounts.
+
+        Args:
+            leapfrog_stepcounts: list of leapfrog stepcounts
+
+        Raises:
+            TypeError: leapfrog stepcounts is not a list
+            ValueError: leapfrog stepcount list is of incorrect length
+            TypeError: leapfrog stepcounts list contains non-integer steps
+            ValueError: leapfrog stepcounts list contains non-positive steps
+
+        Returns:
+            list of validated leapfrog stepcounts
```

Comment:
I return the input from functions because there is some dependency i.e. we can only validate the lengths of leapfrog_stepcounts and leapfrog_stepsizes is correct after validating that the maximum number of proposals is valid itself. Thoughts?